File: java-study/java-syntax/src/main/java/algorithm/LongestSubstringWithoutRepeatingChars.java
package algorithm;

import java.util.*;

/*
 *  Sliding window technique.
 *
 *  Example 1: abcabcbb
 *  Substring: abc
 *  Max length: 3
 *
 *  Example 2: pwwkew
 *  Substring: wke
 *  Subsequence: pwke
 *  Max length: 3
 */
public class LongestSubstringWithoutRepeatingChars {
    public static void main(String[] args) {
        System.out.println(doByKeyValue("abcabcbb"));
        System.out.println(doByPointer("pwwkew"));
    }

    public static int doByKeyValue(String value) {
        int maxLength = 0;
        int leftPointer = 0;
        HashMap<Character, Integer> chars = new HashMap<>();

        for (int rightPointer = 0; rightPointer < value.length(); rightPointer++) {
            char currentChar = value.charAt(rightPointer);
            if (chars.containsKey(currentChar)) {
                leftPointer = Math.max(leftPointer, chars.get(currentChar) + 1);
            }
            chars.put(currentChar, rightPointer);
            int total = rightPointer - leftPointer + 1;
            maxLength = Math.max(maxLength, total);
        }

        return maxLength;
    }

    public static int doByPointer(String value) {
        int leftPointer = 0;
        int rightPointer = 0;
        int maxLength = 0;
        HashSet<Character> chars = new HashSet<>();

        while (rightPointer < value.length()) {
            if (!chars.contains(value.charAt(rightPointer))) {
                chars.add(value.charAt(rightPointer));
                rightPointer++;
                maxLength = Math.max(chars.size(), maxLength);
            } else {
                chars.remove(value.charAt(leftPointer));
                leftPointer++;
            }
        }

        return maxLength;
    }
}

File: kotlin-study/kotlin-syntax/src/main/kotlin/algorithm/LongestSubstringWithoutRepeatingChars.kt
package algorithm

fun main() {
    val algo = LongestSubstringWithoutRepeatingChars()
    println(algo.doByKeyValue("abcabcbb"))
    println(algo.doByPointer("pwwkew"))
}

/*
 *  Sliding window technique.
 *
 *  Example 1: abcabcbb
 *  Substring: abc
 *  Max length: 3
 *
 *  Example 2: pwwkew
 *  Substring: wke
 *  Subsequence: pwke
 *  Max length: 3
 */
class LongestSubstringWithoutRepeatingChars {

    fun doByKeyValue(value: String) : Int {
        var maxLength = 0
        var leftPointer = 0
        val chars = hashMapOf<Char, Int>()

        for (rightPointer in value.indices) {
            val currentChar = value[rightPointer]
            if (chars.containsKey(currentChar)) {
                leftPointer = leftPointer.coerceAtLeast(chars[currentChar]!! + 1) // max
            }
            chars[currentChar] = rightPointer
            val total = rightPointer - leftPointer + 1
            maxLength = maxLength.coerceAtLeast(total) // max
        }

        return maxLength
    }

    fun doByPointer(value: String) : Int {
        var leftPoint = 0
        var rightPoint = 0
        var maxLength = 0
        val chars = hashSetOf<Char>()

        while (rightPoint < value.length) {
            if (!chars.contains(value[rightPoint])) {
                chars.add(value[rightPoint])
                rightPoint++
                maxLength = chars.size.coerceAtLeast(maxLength) // max
            } else {
                chars.remove(value[leftPoint])
                leftPoint++
            }
        }

        return maxLength
    }
}
